R/assignment_linter.R

```diff
@@ -2,27 +2,27 @@
 #'
 #' Check that `<-` is always used for assignment.
 #'
-#' @param block_double_assign Logical, `FALSE` by default. If `TRUE`, usage
-#'   of [`<<-`][base::assignOps] and [`->>`][base::assignOps] are also blocked.
+#' @param allow_cascading_assign Logical, default `TRUE`.
+#'   If `FALSE`, [`<<-`][base::assignOps] and `->>` are not allowed.
+#' @param allow_right_assign Logical, default `FALSE`. If `TRUE`, `->` and `->>` are allowed.
 #' @evalRd rd_tags("assignment_linter")
 #' @seealso [linters] for a complete list of linters available in lintr.
 #' @export
-assignment_linter <- function(block_double_assign = FALSE) {
+assignment_linter <- function(allow_cascading_assign = TRUE, allow_right_assign = FALSE) {
   Linter(function(source_file) {
     if (is.null(source_file$xml_parsed_content)) return(list())
 
     xml <- source_file$xml_parsed_content
 
-    node_conditions <- paste0("self::", c(
-      # always block = (NB: the parser differentiates EQ_ASSIGN and EQ_SUB)
-      "EQ_ASSIGN",
-      # -> and ->> are both 'RIGHT_ASSIGN'; check the text if not blocking ->>
-      paste0("RIGHT_ASSIGN", if (!block_double_assign) "[text() = '->']"),
+    # TODO: is there any performance consideration for //*[self::a or self::b] vs. //a|//b ?
+    xpath <- paste(collapse = " | ", c(
+      # always block = (NB: the parser differentiates EQ_ASSIGN, EQ_SUB, and EQ_FORMALS)
+      "//EQ_ASSIGN",
+      # -> and ->> are both 'RIGHT_ASSIGN'
+      if (!allow_right_assign) "//RIGHT_ASSIGN" else if (!allow_cascading_assign) "//RIGHT_ASSIGN[text() = '->>']",
       # <-, :=, and <<- are all 'LEFT_ASSIGN'; check the text if blocking <<-
```

Comment:
Shouldn't we lint `:=` as well?

Reply:
We'd have to write special logic to handle correct `rlang` and `data.table` usages. Don't think it's worth it (most people don't know that `:=` can be used for "regular" assignment and it's explicitly not mentioned in the R documentation).

Reply:
Agreed. Maybe worth mentioning here why we chose not to lint `:=`?